syntax = "proto3";

package seigr.access_control;

// ===========================
// ENUMS
// ===========================

/**
 * Represents different user role types within the access control system.
 * Roles define a user's level of access and responsibilities.
 */
enum RoleType {
    ROLE_TYPE_UNDEFINED = 0;       // Undefined or unknown role.
    ROLE_TYPE_ADMIN = 1;           // Full administrative access.
    ROLE_TYPE_EDITOR = 2;          // Can edit resources but limited administrative control.
    ROLE_TYPE_VIEWER = 3;          // Read-only access.
    ROLE_TYPE_CONTRIBUTOR = 4;     // Can contribute content but with restrictions.
    ROLE_TYPE_AUDITOR = 5;         // Can only view audit logs and reports.
    ROLE_TYPE_DELEGATOR = 6;       // Can delegate permissions to other users.
    ROLE_TYPE_API = 7;             // Role for API-based access.
    ROLE_TYPE_SYSTEM = 8;          // Reserved for internal system processes.
}

/**
 * Defines granular permission types available for resources and actions.
 */
enum PermissionType {
    PERMISSION_TYPE_UNDEFINED = 0; // Undefined or unknown permission.
    PERMISSION_TYPE_READ = 1;      // Permission to read data.
    PERMISSION_TYPE_WRITE = 2;     // Permission to modify data.
    PERMISSION_TYPE_EXECUTE = 3;   // Permission to execute operations.
    PERMISSION_TYPE_DELETE = 4;    // Permission to delete resources.
    PERMISSION_TYPE_SHARE = 5;     // Permission to share resources.
    PERMISSION_TYPE_MANAGE = 6;    // Permission to manage permissions.
    PERMISSION_TYPE_DELEGATE = 7;  // Permission to delegate access.
    PERMISSION_TYPE_AUDIT = 8;     // Permission to audit logs.
    PERMISSION_TYPE_BYPASS = 9;    // Critical bypass permission.
}

/**
 * Represents types of access actions performed on resources.
 */
enum AccessType {
    ACCESS_TYPE_UNDEFINED = 0;     // Undefined access type.
    ACCESS_TYPE_READ = 1;          // Read access.
    ACCESS_TYPE_WRITE = 2;         // Write access.
    ACCESS_TYPE_EXECUTE = 3;       // Execute access.
    ACCESS_TYPE_DELETE = 4;        // Delete access.
    ACCESS_TYPE_SHARE = 5;         // Share access.
    ACCESS_TYPE_ADMINISTRATE = 6;  // Administrative access.
    ACCESS_TYPE_DELEGATE = 7;      // Access delegation.
    ACCESS_TYPE_API_CALL = 8;      // API-based access.
}

/**
 * Represents the status of an access control policy.
 */
enum PolicyStatus {
    POLICY_STATUS_UNDEFINED = 0;       // Undefined policy status.
    POLICY_STATUS_ACTIVE = 1;          // Policy is active and enforced.
    POLICY_STATUS_INACTIVE = 2;        // Policy is inactive.
    POLICY_STATUS_EXPIRED = 3;         // Policy has expired.
    POLICY_STATUS_REVOKED = 4;         // Policy was revoked.
    POLICY_STATUS_PENDING_REVIEW = 5;  // Policy is pending review.
    POLICY_STATUS_SCHEDULED_REVIEW = 6;// Policy is scheduled for review.
}

/**
 * Represents the result of an access request evaluation.
 */
enum AccessDecisionType {
    ACCESS_DECISION_UNDEFINED = 0; // Undefined decision.
    ACCESS_DECISION_ALLOW = 1;     // Access is allowed.
    ACCESS_DECISION_DENY = 2;      // Access is denied.
    ACCESS_DECISION_CONDITIONAL = 3; // Access allowed with conditions.
    ACCESS_DECISION_PENDING = 4;   // Decision is pending evaluation.
}

// ===========================
// MESSAGES
// ===========================

/**
 * Represents a user role with permissions and constraints.
 */
message Role {
    string role_id = 1;                          // Unique identifier for the role.
    string name = 2;                             // Descriptive name of the role.
    repeated PermissionType permissions = 3;     // List of permissions associated with the role.
    map<string, string> constraints = 4;         // Contextual constraints (e.g., time, location).
    bool can_delegate = 5;                       // Whether the role can delegate permissions.
    int32 inheritance_depth = 6;                 // Level of inherited permissions.
    repeated string parent_roles = 7;            // Parent roles for inheritance.
}

/**
 * Defines permission settings for a specific user or entity.
 */
message Permission {
    string entity_id = 1;                        // ID of the user or entity.
    RoleType role = 2;                           // Role associated with the entity.
    repeated PermissionType actions = 3;         // Actions the entity is allowed to perform.
    string constraints = 4;                      // Time or location-based constraints.
    string expiry = 5;                           // Expiration date for temporary permissions.
    bool requires_mfa = 6;                       // Whether MFA is required for access.
    map<string, string> metadata = 7;            // Additional metadata.
    bool conditional_access = 8;                 // Conditional access flag.
}

/**
 * Defines an access control entry for a user or entity.
 */
message AccessControlEntry {
    string user_id = 1;                          // ID of the user/entity.
    RoleType role = 2;                           // Assigned role.
    repeated PermissionType permissions = 3;     // List of granted permissions.
    repeated string allowed_actions = 4;         // Specific allowed actions.
    bool is_inherited = 5;                       // Whether permissions are inherited.
    string effective_from = 6;                   // Start date for permissions.
    string expires_on = 7;                       // Expiration date for permissions.
    bool mfa_required = 8;                       // MFA requirement flag.
    string delegated_by = 9;                     // ID of the user who delegated access.
    string access_reason = 10;                   // Reason for granting access.
}

/**
 * Defines an access control list (ACL) containing multiple entries.
 */
message AccessControlList {
    repeated AccessControlEntry entries = 1;     // List of access control entries.
    string version = 2;                          // ACL version.
    map<string, string> metadata = 3;            // Metadata for tracking changes.
    string policy_id = 4;                        // Governing policy ID.
    PolicyStatus status = 5;                     // Status of the ACL.
    string last_reviewed = 6;                    // Timestamp of last review.
    string next_review_due = 7;                  // Next scheduled review date.
}

/**
 * Represents an access control policy.
 */
message AccessPolicy {
    string policy_id = 1;                        // Unique identifier for the policy.
    string policy_name = 2;                      // Human-readable policy name.
    PolicyStatus status = 3;                     // Current status of the policy.
    string creation_timestamp = 4;               // Policy creation timestamp.
    string updated_timestamp = 5;                // Last updated timestamp.
    repeated Permission permissions = 6;         // List of permissions in the policy.
    string enforced_on = 7;                      // Enforcement date.
    map<string, string> metadata = 8;            // Additional metadata.
    repeated string applicable_roles = 9;        // Roles governed by the policy.
}

/**
 * Logs detailed access-related events for auditing.
 */
message AccessAuditLog {
    string user_id = 1;
    AccessType action = 2;
    string timestamp = 3;
    string status = 4;
    string details = 5;
    string ip_address = 6;
    string device_info = 7;
    bool mfa_verified = 8;
    bool delegated_access = 9;
    string origin_role = 10;
    string access_reason = 11;
}
